expression/builtin_cast.go

```diff
@@ -1840,6 +1840,11 @@ func BuildCastCollationFunction(ctx sessionctx.Context, expr Expression, ec *Exp
 
 // BuildCastFunction builds a CAST ScalarFunction from the Expression.
 func BuildCastFunction(ctx sessionctx.Context, expr Expression, tp *types.FieldType) (res Expression) {
+	argType := expr.GetType()
+	// If source argument's nullable, then target type should be nullable
+	if !mysql.HasNotNullFlag(argType.Flag) {
+		tp.Flag &= ^mysql.NotNullFlag
+	}
```

Comment:
Looks like we have a similar problem in other functions, like coalesce, if... you can double-check.
Prefer to define an interface like "nullable(args []Expression) bool" and force every function to implement it. So that next time we add a new function, we won't forget to consider the null handling.

Reply:
Nice comments, I'll discuss it with @fzhedu.